DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/buffered_host_streaming/src/common.hpp

```diff
@@ -0,0 +1,110 @@
+#ifndef __COMMON_HPP__
+#define __COMMON_HPP__
+
+#include <algorithm>
+#include <chrono>
+#include <numeric>
+#include <vector>
+#include <thread>
+
+// A multithreaded version of memcpy.
+// On modern processors with multiple cores and threads, a single-threaded
+// memcpy cannot saturate the memory bandwidth (not even close). Moreover,
+// a single-threaded memcpy doesn't even get close to the PCIe bandwidth.

```

Comment:
This is surprising. Even if `memcpy` is vectorized? I guess with AVX512 you have 64B/cycle, so at 4 GHz this is 256 GB/s.
Or perhaps you should use processors from other vendors? ;-)

Reply:
I did quite a bit of experimenting and discussing to convince myself of this :|.

1) Max bandwidth for the (super awesome, can't be beat) CPU I ran on is 68 GB/s ([source](https://ark.intel.com/content/www/us/en/ark/products/82765/intel-xeon-processor-e5-1650-v3-15m-cache-3-50-ghz.html)). Memcpy is read+write so max throughput is 34 GB/s. I am running two memcpys in parallel (producer and consumer) so max throughput is 17 GB/s.

2) The key is the latency from L1 data cache to the memory system. If that latency is high, you need more memory requests interleaved in flight to actually see a cumulative bandwidth that maxes out the theoretical bandwidth. In the case of server-grade CPUs this latency is significantly higher than non-server CPUs. [Here ](https://stackoverflow.com/questions/39260020/why-is-skylake-so-much-better-than-broadwell-e-for-single-threaded-memory-throug) is a more deep discussion.

3) I measured std::copy to be noticeably worse than memcpy. I thought, it would be equivilant (and just fall back to memcpy), and maybe in some cases win  since it has type information.

4) What do you suggest? Maybe a Ryzen ;)?

Reply:
What an old processor. :-) Interesting feedback. Could be added somehow as a design comment in the example? Perhaps non temporal memory accesses would help?
Otherwise, I did not dare to suggest using some other vendor's processor. I have not tried them yet... :-)
But since you mentioned it yourself, you are fired! ;-)